Add Psi4 to GeneralRelativity

AlexCarpenter46 opened the pull request:
## Proposed changes

This pr is split up into 2 commits, one to add the files and functions needed for Psi4 and one to add the ability to observe the real portion of Psi4 in BBH and Kerrschild.


### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

I didn't include the imaginary portion in output, can if we would like that functionality as well.

geoffrey4444 commented on tests/Unit/PointwiseFunctions/GeneralRelativity/CMakeLists.txt:
```diff
@@ -12,6 +12,7 @@ set(LIBRARY_SOURCES
   Test_KerrSchildCoords.cpp
   Test_ProjectionOperators.cpp
   Test_Ricci.cpp
+  Test_Psi4.cpp
```
alphabetize

geoffrey4444 commented on src/PointwiseFunctions/GeneralRelativity/Tags.hpp:
```diff
@@ -152,6 +152,14 @@ struct SpatialRicciScalar : db::SimpleTag {
 };
 
+/*!
+ * \brief Computes the real portion of \f$\Psi_4\f$
```
I would say real part, not real portion

geoffrey4444 commented on src/PointwiseFunctions/GeneralRelativity/Psi4Real.hpp:
```diff
@@ -0,0 +1,94 @@
+    const tnsr::ii<RealDataType, 3, Frame>& spatial_metric,
+    const tnsr::II<RealDataType, 3, Frame>& inverse_spatial_metric,
+    const tnsr::I<RealDataType, 3, Frame>& inertial_coords) {
+  const auto psi_4_complex = psi_4<ComplexDataVector>(
```
instead of storing psi_4_complex in a variable, just put all this inside the `get()` on the RHS of line 44

geoffrey4444 commented on src/PointwiseFunctions/GeneralRelativity/Psi4Real.hpp:
```diff
@@ -0,0 +1,94 @@
+    const tnsr::ii<RealDataType, 3, Frame>& spatial_metric,
+    const tnsr::II<RealDataType, 3, Frame>& inverse_spatial_metric,
+    const tnsr::I<RealDataType, 3, Frame>& inertial_coords) {
+  auto psi_4_real_result = make_with_value<Scalar<RealDataType>>(
```
Initialize with signaling nan instead of zero

geoffrey4444 commented on src/PointwiseFunctions/GeneralRelativity/Psi4Real.hpp:
```diff
@@ -0,0 +1,94 @@
+
+
+namespace Tags {
+/// Computes the real portion of the Newman Penrose quantity \f$\Psi_4\f$ using
```
I would say real part, not real portion

AlexCarpenter46 replied:
forgot to put this change in my fixup, will put it in for the next comments I get

geoffrey4444 commented on src/PointwiseFunctions/GeneralRelativity/Psi4Real.hpp:
```diff
@@ -0,0 +1,94 @@
+ * \brief Computes the real portion of the Newman Penrose quantity \f$\Psi_4\f$
+ * using  \f$\Psi_4[Real] = -0.5*U^{8+}_{ij}*(x^ix^j - y^iy^j)\f$.
+ */
+template <typename Frame, typename RealDataType>
```
I would put the definition of `psi_4_real` in `Psi4Real.cpp` and only put the declaration in the hpp

geoffrey4444 commented on src/PointwiseFunctions/GeneralRelativity/Psi4.hpp:
```diff
@@ -0,0 +1,61 @@
+ * U\f$^{8+}\f$ = (\f$P^{(a}_i P^{b)}_j\f$ - \f$\frac{1}{2}P_{ij}P^{ab}\f$)
+ * (\f$E_{ab} - \epsilon_a^{cd}n_dB_{cb}\f$)
+ * and \f$\bar{m}^i\f$ = \f$\frac{(x^i + iy^i)}{\sqrt{2}}\f$. \f$x^i and y^i\f$
+ * are normalized unit vectors in the inertial frame.
```
you template on frame, so shouldn't these be unit vectors in frame Frame?

geoffrey4444 commented on src/PointwiseFunctions/GeneralRelativity/Psi4.cpp:
```diff
@@ -0,0 +1,143 @@
+  auto x_coord = make_with_value<tnsr::I<RealDataType, 3, Frame>>(
+      get<0, 0>(inverse_spatial_metric), 0.0);
+  x_coord.get(0) = 1.0;
+  const auto magnitude_x =
```
no need to allocate magnitude_x or x_coord here...just put these right into line 71, and similarly for y_hat

geoffrey4444 commented on src/PointwiseFunctions/GeneralRelativity/Psi4.cpp:
```diff
@@ -0,0 +1,143 @@
+    const tnsr::II<RealDataType, 3, Frame>& inverse_spatial_metric,
+    const tnsr::I<RealDataType, 3, Frame>& inertial_coords) {
+  auto psi_4_result = make_with_value<Scalar<ComplexDataType>>(
+      get<0, 0>(inverse_spatial_metric), 0.0);
```
initialize with signaling nan not zero

geoffrey4444 commented on src/PointwiseFunctions/GeneralRelativity/Psi4.cpp:
```diff
@@ -0,0 +1,143 @@
+#define CDTYPE(data) BOOST_PP_TUPLE_ELEM(1, data)
+#define FRAME(data) BOOST_PP_TUPLE_ELEM(2, data)
+
+#define INSTANTIATE(_, data)                                                 \
```
clang format

geoffrey4444 commented on src/PointwiseFunctions/GeneralRelativity/Psi4.cpp:
```diff
@@ -0,0 +1,143 @@
+    } else {
+      if (magnitude_cartesian.get() != 0.0) {
+        r_hat.get(j) = inertial_coords.get(j) / magnitude_cartesian.get();
+      }
```
These warnings are saying you never test this code. If these persist after my suggested changes, we'll have to figure out why your test isn't covering these lines

geoffrey4444 commented on tests/Unit/PointwiseFunctions/GeneralRelativity/Test_Psi4.cpp:
```diff
@@ -0,0 +1,74 @@
+                  "[Unit][PointwiseFunctions]") {
+  pypp::SetupLocalPythonEnvironment local_python_env("PointwiseFunctions/");
+
+  const size_t size = 1;
```
why not test for size 2? Have a couple different grid points

geoffrey4444 commented on tests/Unit/PointwiseFunctions/GeneralRelativity/Test_Psi4.cpp:
```diff
@@ -0,0 +1,74 @@
+  const auto expected = gr::psi_4<ComplexDataType>(
+      spatial_ricci, extrinsic_curvature, cov_deriv_extrinsic_curvature,
+      spatial_metric, inv_spatial_metric, inertial_coords);
+  CHECK_ITERABLE_APPROX(expected, python_psi_4);
```
also test the real part is computed correctly

geoffrey4444 reviewed (changes requested):
Nice start! Some suggested changes. We also have to figure out the codecov stuff

nilsdeppe commented on src/PointwiseFunctions/GeneralRelativity/Psi4.hpp:
```diff
@@ -0,0 +1,61 @@
+/// @{
+/*!
+ * \ingroup GeneralRelativityGroup
+ * \brief Computes Newman Penrose quantity $\Psi_4$ using the characteristic
```
`$` -> `\f$`

nilsdeppe commented on src/PointwiseFunctions/GeneralRelativity/Psi4.hpp:
```diff
@@ -0,0 +1,61 @@
+ * \brief Computes Newman Penrose quantity $\Psi_4$ using the characteristic
+ * field U\f$^{8+}\f$ and complex vector \f$\bar{m}^i\f$.
+ *
+ * \details Computes $\Psi_4$ as: \f$\Psi_4\f$ =
```
Why jump in and out of math mode here instead of doing the entire equation in math mode?

nilsdeppe commented on src/PointwiseFunctions/GeneralRelativity/Psi4.hpp:
```diff
@@ -0,0 +1,61 @@
+ * U\f$^{8+}_{ij}\bar{m}^i\bar{m}^j\f$ with the characteristic field
+ * U\f$^{8+}\f$ = (\f$P^{(a}_i P^{b)}_j\f$ - \f$\frac{1}{2}P_{ij}P^{ab}\f$)
+ * (\f$E_{ab} - \epsilon_a^{cd}n_dB_{cb}\f$)
+ * and \f$\bar{m}^i\f$ = \f$\frac{(x^i + iy^i)}{\sqrt{2}}\f$. \f$x^i and y^i\f$
```
for `x^i and y^` you shuld jump out of math mode for `and`

nilsdeppe commented on src/PointwiseFunctions/GeneralRelativity/Psi4.cpp:
```diff
@@ -0,0 +1,137 @@
+
+namespace gr {
+template <typename ComplexDataType, typename Frame, typename RealDataType>
+void psi_4(
```
I'm a bit worried about the number of memory allocations in this function. It's pretty huge. Do we even need a `double` variant or is `DataVector`-only one enough? If the latter is enough, I'd say assume DataVector and then use a `Variables` to combine the allocations and using the `not_null` calls to functions.

I also don't understand why it says a bunch of lines are untested, though it codecov tends to be fairly accurate. My guess is because you aren't testing the `double` case. Since it's not tested and I don't think we will need it, I say just hard-code DV and CDV :)

AlexCarpenter46 replied:
I think DataVector should be fine, I'll switch that over soon. 

Also side note, was working on testing doubles and found that pypp:call doesn't support std::complex<double> so testing doubles would've been a hassle.

nilsdeppe replied:
Oh, yea, that could be!

nilsdeppe commented on src/PointwiseFunctions/GeneralRelativity/Psi4.cpp:
```diff
@@ -0,0 +1,137 @@
+      get<0, 0>(spatial_metric), 0.0);
+  for (size_t j = 0; j < 3; j++) {
+    if constexpr (is_derived_of_vector_impl_v<RealDataType>) {
+      for (size_t i = 0; i < get(magnitude_cartesian).size(); i++) {
```
You don't need the `if-constexpr` you can use `get_size` and `get_element` from `Utilities/ContainerHelpers.hpp`

AlexCarpenter46 replied:
If we're not going to support doubles, then I think it's safe to take the if statement out altogether.

nilsdeppe replied:
Yep, agreed!

nilsdeppe reviewed (changes requested):
Overall LGTM, just my usual "let's elide memory allocations" comments 🙃  You can squash, my changes will touch enough lines that I'll need to think about the code anyway :)

nilsdeppe commented on src/PointwiseFunctions/GeneralRelativity/Psi4.cpp:
```diff
@@ -0,0 +1,144 @@
+  const std::complex<double> imag = std::complex<double>(0.0, 1.0);
+  auto& y_hat =
+      get<::Tags::TempI<0, 3, Frame, ComplexDataVector>>(y_hat_buffer);
+  // std::complex<double> + DataVector -> ComplexDataVector
```
Sorry, what is this note for? The next line doesn't do any addition. It might just be left over from before and can be deleted, or maybe it needs to be changed to `*`?

AlexCarpenter46 replied:
Oh whoops, yeah that should be a * this was when Alex Macedo and I were trying to figure out if tenex should handle a complex * real = complex. It's probably not needed anymore

nilsdeppe reviewed (dismissed):
Looks great! Excellent work, especially with the Variables changes! Those are not always the easiest to understand :D

It looks like there's a conflict (probably my fault for merging some other PR).

Could you please rebase to resolve the conflict and then ping me so I can get this merged before you get more conflicts :D

nilsdeppe commented on tests/Unit/PointwiseFunctions/GeneralRelativity/Test_Psi4.cpp:
```diff
@@ -0,0 +1,122 @@
+      determinant_and_inverse(spatial_metric).second;
+  const auto inertial_coords =
+      make_with_random_values<tnsr::I<RealDataType, 3, Frame::Inertial>>(
+          nn_generator, nn_distribution, used_for_size_real);
```
Can you explicitly set one of the coordinates to r=0? E.g.
```cpp
for (size_t i = 0; i < 3; ++i) {
  inertial_coords.get(i)[0] = 0.0;
}
```
The test coverage is showing you aren't hitting the r=0 branch (because random numbers)

nilsdeppe commented on tests/Unit/PointwiseFunctions/GeneralRelativity/Test_Psi4.cpp:
```diff
@@ -0,0 +1,122 @@
+                  "[Unit][PointwiseFunctions]") {
+  pypp::SetupLocalPythonEnvironment local_python_env("PointwiseFunctions/");
+
+  const size_t size = 2;
```
We typically use 5

nilsdeppe reviewed (commented):
Sorry, couple things I noticed from code coverage

AlexCarpenter46 committed:
Add Psi4 to GeneralRelativity

AlexCarpenter46 committed:
Add ability to observe real part of Psi4

AlexCarpenter46 commented:
Added those changes and noticed that my python test wasn't handling the 0 case correctly, einsum formats things strangely. Fixed it up and made the python test look more like the cpp with the correct names and all!
